### PR TITLE
Additional Steps-Certificate Tar file when Upgrade

### PR DESCRIPTION
**Pull request by dubewarsagar**
Added 2 additional steps while upgrading {smart-proxy-context} server
Step-2:Confirm cert tar file on existing {smart-proxy-context} server
Step-4:If you use custom SSL certificates on....

A reference to Installing {smart-proxy-context} Server guide is
moved from step 3 to 4.

https://bugzilla.redhat.com/show_bug.cgi?id=2104948

* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [x] Foreman 3.3/Katello 4.5
* [x] Foreman 3.2/Katello 4.4
* [x] Foreman 3.1/Katello 4.3
* For Foreman 3.0 or older, please create a separate PR.
* We do not accept PRs for Foreman 2.3 or older.

### CODE REVIEW

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:53]:
```suggestion
Use the same name `myproxy.example.com-certs.tar` to regenerate certificates on {ProjectServer}.
```

See https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L81

We should also substitue the "myproxy.example.com" FQDN with an attribute.

Please also reword "should be used" to "use".

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:49]: It's most likely named "katello-answers" or "foreman-answers" on other scenarios; so we have to use an attribute.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:66]:
```suggestion
. If you use custom SSL certificates on your {smart-proxy-context} server, use the following command to regenerate the certificate tar file on {ProjectServer}:
```

Can we use "SmartProxyServer" instead of "{smart-proxy-context} server"?

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:75]:
```suggestion
# {smart-proxy-context}-certs-generate \
--foreman-proxy-fqdn "myproxy.example.com" \
--certs-tar "~/myproxy.example.com-certs.tar" \
--server-cert "/root/{smart-proxy-context}_cert/{smart-proxy-context}_cert.pem" \
--server-key "/root/{smart-proxy-context}_cert/{smart-proxy-context}_cert_key.pem" \
--server-ca-cert "/root/{smart-proxy-context}_cert/ca_cert_bundle.pem" \
--certs-update-server
```

Please also use attributes instead of "myproxy".

**Review by maximiliankolb — CHANGES_REQUESTED** [review]

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:53]: > See https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L81
> 
> We should also substitue the "myproxy.example.com" FQDN with an attribute.
No idea which attribute should be used here. I checked  all the attribute files, but I did not get any attribute for myproxy. 
> 
> Please also reword "should be used" to "use".
I think this is no longer required after changing sentence to what you suggested earlier. The final sentence sounds like:
. Use the same name `myproxy.example.com-certs.tar` to regenerate certificates on {ProjectServer}.
Kindly inform if I am missing something.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:66]: Will do.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:75]: same as above, 
No idea which attribute should be used here. I checked  all the attribute files, but I did not get any attribute for myproxy.

Comment (maximiliankolb) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:75]: You can use "smartproxy-example-com":

```
guides/common/attributes-base.adoc:94::smartproxy-example-com: smartproxy.example.com
guides/common/attributes-orcharhino.adoc:44::smartproxy-example-com: orcharhino-proxy.example.com
guides/common/attributes-satellite.adoc:85::smartproxy-example-com: capsule.example.com
```

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (ekohl) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:49]: As an installer maintainer I really don't like grepping in the answers file in our instructions. Answer files should be considered an internal installer detail.

IMHO you should always pass in the value if you want to guarantee it's set to a certain value, rather than making sure files are in places that the answers dictate.

So in practice:
* Generate the tar file on the `{Project} server`
* Copy it over to `{smart-proxy-context} server`
* Run the installer with `--tar-file /path/to/tar` on `{smart-proxy-context} server`.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:75]: There are four more instances of **myproxy.example.com**, shall we replace them with **smartproxy-example-com** too? 
I am asking this as the original file (procedure) not written by me.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:49]: Not sure what changes are required in the doc.

Comment (ekohl) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:49]: Please drop this step since it's not needed. All the way at the bottom the installer command is:
```
# foreman-installer --certs-tar-file /root/myproxy.example.com-certs.tar \
```

That means the actual answer for tar_file in `*-answers.yaml` is set to `/root/myproxy.example.com-certs.tar`, regardless of what the value was. Checking it at this point can only create confusion.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (ekohl) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:53]: Please also drop this line.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:75]: Yes, that would be much appreciated.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (ekohl) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:53]: I think this should be dropped. Why is it here?

**Review by ekohl — CHANGES_REQUESTED** [review]

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc]: Dropped this step and added a new step after the discussion with original issue created. Thanks :)

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc]: Removed.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (ekohl) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:64]: AFAIK these are user provided paths. Perhaps you could be more explicit in that they're user provided paths?
```suggestion
--server-cert "/path/to/{smart-proxy-context}_cert.pem" \
--server-key "/path/to/{smart-proxy-context}_key.pem" \
--server-ca-cert "/path/to/{smart-proxy-context}_ca.pem" \
```

**Review by ekohl — COMMENTED** [review]
I wonder why the preview could not be generated. I'll look into it because it makes reviewing easier.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (ekohl) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:59]: Now that the preview renders, I see this is incorrect:
```suggestion
# {certs-generate} \
```

**Review by ekohl — CHANGES_REQUESTED** [review]

**Review by ekohl — COMMENTED** [review]
The content reads ok, but could you squash the commits and properly summarize the change in the commit message? Right now the commit and PR don't describe at all what and why you're changing.

**Comment by dubewarsagar** [comment]
> The content reads ok, but could you squash the commits and properly summarize the change in the commit message? Right now the commit and PR don't describe at all what and why you're changing.

On it.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (ekohl) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:78]: I now further read this and this step is redundant, at least in the Katello case. You already run it lower, after `yum -y update`.

I'm not sure what happens in the downstream case. There it may be needed, or place it in the file the answers already exists.

Apologies for the confusion and back and forth on this, but I think this is really the result of a very vague PR description. You don't explain what you are doing and why. That means as a reviewer I'm having to guess and that causes problems. That is why we always insist on clear descriptions of PRs. It makes the review process easier.

**Review by ekohl — COMMENTED** [review]
The whole "Additional Steps" is vague to me. I would use:
> Document custom SSL certificates during upgrade

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:78]: > I now further read this and this step is redundant, at least in the Katello case. You already run it lower, after `yum -y update`.
> 
> I'm not sure what happens in the downstream case. There it may be needed, or place it in the file the answers already exists.
> 
> Apologies for the confusion and back and forth on this, but I think this is really the result of a very vague PR description. You don't explain what you are doing and why. That means as a reviewer I'm having to guess and that causes problems. That is why we always insist on clear descriptions of PRs. It makes the review process easier.

Apologies for the confusion in PR msg, but kindly make it clear what changes do you suggest now. Should we remove this step or any other update is needed? It would have been very easier if this gets clear in the BZ itself. We have changed it a lot since we made the original commit as per the discussion in the BZ mentioned in the original commit msg. 

Thanks!

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:70]:
```suggestion
. Copy the resulting tarball to your {SmartProxy}:
```

If you want, you can either use `_My_{smartproxy-example-com}-certs.tar_` or add `Adjust the path to the tarball.` below the code block (with a `+` above).

Using "we" is discouraged.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:78]: @ekohl, gentle reminder!

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:53]:
```suggestion
```

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:54]:
```suggestion
```

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (maximiliankolb) [inline guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:64]: The single underscore sounds off to me. Maybe set the whole path in italic, i.e. adding another `_` right after `"`? Please make it consistent across all parameters and all of your changes.

File: guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc
Comment (ekohl) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:78]: > It would have been very easier if this gets clear in the BZ itself.

This is a much deeper issue, but I think the problem here is that the experts on the matter are not looking at the BZs. Even worse, the BZ is full of private comments. It is a very downstream focused way of working. In the ideal world all a reviewer needs is this single PR and has all the information. As the PR author that is your job.

Because I thought it was easier to demonstrate using code, I'd have expected something like https://github.com/theforeman/foreman-documentation/pull/1499. That has the context of why it splits the instructions. I also think it is the most minimal solution to the problem, though the phrasing can probably be improved.

Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:78]: > > It would have been very easier if this gets clear in the BZ itself.
> 
> This is a much deeper issue, but I think the problem here is that the experts on the matter are not looking at the BZs. Even worse, the BZ is full of private comments. It is a very downstream focused way of working. In the ideal world all a reviewer needs is this single PR and has all the information. As the PR author that is your job.

The BZ has private comments and there is no essential information in it which would hamper the working of upstream. The private comments are added for reason and for internal use only. In most cases, they are requests and information on reassigning tasks. 
 
> Because I thought it was easier to demonstrate using code, I'd have expected something like #1499. That has the context of why it splits the instructions. I also think it is the most minimal solution to the problem, though the phrasing can probably be improved.

What I have observed is approach has been changed over a period of time spent in working of this BZ. Which the writer absolutely has no control over. Also, updates are being suggested for the part which is not added or updated as a part of this PR. This makes the PR more complected. 

If PR #1499 having all the changes in the way it is expected, kindly inform if this PR can be closed without merge and update the BZ accordingly.

Comment (ekohl) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:78]: > What I have observed in this BZ is, approach has been changed over a period of time spent in working of this BZ. Which the writer absolutely has no control over. Also, updates are being suggested for the part which is not added or updated as a part of this PR. This makes the PR more complected.

That's true, mostly because the BZ also didn't consider that upstream and downstream have different procedures, leading to more confusion. The described problem only exists in the downstream procedure while I was looking at it and seeing you modify the upstream procedure.

And I do agree that the scope was being increased. However, it's perfectly fine to postpone that work to a later PR. That's also what I did in #1499. Just because someone asks something, doesn't mean you have to do it. It's always a discussion.

To be more constructive: if in the future anything regarding the installer or platform is unclear, please ask me or anyone else from the platform team. I'd be happy to take a look early in the process. Hopefully we can remove a lot of the back and forth then.

> Suggestions are not provided all at once. They are provided in parts. This makes the PR more complected.

That's true. As I said, that often comes from an unclear problem description in the PR so it's hard to verify you're actually achieving what you intend to.

> If PR https://github.com/theforeman/foreman-documentation/pull/1499 having all the changes in the way it is expected, kindly inform if this PR can be closed without merge and update the BZ accordingly.

I think it is. On a technical level I think it solves the issue, but I'm not a writer so I'd appreciate a critical eye on that.

Comment (dubewarsagar) [reply guides/doc-Upgrading_and_Updating/topics/upgrading_capsule_server.adoc:78]: > > Suggestions are not provided all at once. They are provided in parts. This makes the PR more complected.
> 
> That's true. As I said, that often comes from an unclear problem description in the PR so it's hard to verify you're actually achieving what you intend to.
Unclear PR description does not mean you should review only half of the content and leave rest of the part. :)
> 
> > If PR #1499 having all the changes in the way it is expected, kindly inform if this PR can be closed without merge and update the BZ accordingly.
> 
> I think it is. On a technical level I think it solves the issue, but I'm not a writer so I'd appreciate a critical eye on that.
Thank you for confirming that this PR can be closed. I am closing this one and you may continue merging the other PR when its ready.

**Comment by dubewarsagar** [comment]
Closed withut merge. For further updates, please see PR:  https://github.com/theforeman/foreman-documentation/pull/1499

**Pull request closed without merging**
